lib/segment/src/index/sparse_index/sparse_vector_index.rs

```diff
@@ -142,91 +93,154 @@ impl<TInvertedIndex: InvertedIndex> SparseVectorIndex<TInvertedIndex> {
         unique_record_ids.len()
     }
 
-    /// Plain search not using the inverted index.
-    pub fn search_plain(
+    fn get_query_cardinality(&self, filter: &Filter) -> CardinalityEstimation {
+        let vector_storage = self.vector_storage.borrow();
+        let id_tracker = self.id_tracker.borrow();
+        let payload_index = self.payload_index.borrow();
+        let available_vector_count = vector_storage.available_vector_count();
+        let query_point_cardinality = payload_index.estimate_cardinality(filter);
+        adjust_to_available_vectors(
+            query_point_cardinality,
+            available_vector_count,
+            id_tracker.available_point_count(),
+        )
+    }
+
+    fn search_plain(
         &self,
-        vectors: &[&QueryVector],
-        filter: &Filter,
+        query_vector: &QueryVector,
+        filter: Option<&Filter>,
         top: usize,
         is_stopped: &AtomicBool,
-    ) -> OperationResult<Vec<Vec<ScoredPointOffset>>> {
-        let mut results = Vec::with_capacity(vectors.len());
+    ) -> OperationResult<Vec<ScoredPointOffset>> {
+        let vector_storage = self.vector_storage.borrow();
         let id_tracker = self.id_tracker.borrow();
-        let payload_index = self.payload_index.borrow();
-        let vector_storage = &self.vector_storage.borrow();
-        // run filter once for all vectors
-        let filtered_points = payload_index.query_points(filter);
-        for &vector in vectors {
-            check_process_stopped(is_stopped)?;
-            let _timer = ScopeDurationMeasurer::new(&self.searches_telemetry.unfiltered_sparse);
-            let raw_scorer = new_stoppable_raw_scorer(
-                vector.clone(),
-                vector_storage,
-                id_tracker.deleted_point_bitslice(),
-                is_stopped,
-            )?;
-            let search_results =
-                raw_scorer.peek_top_iter(&mut filtered_points.iter().copied(), top);
-            results.push(search_results);
+        let raw_scorer = new_stoppable_raw_scorer(
+            query_vector.clone(),
+            &vector_storage,
+            id_tracker.deleted_point_bitslice(),
+            is_stopped,
+        )?;
+        match filter {
+            Some(filter) => {
+                let payload_index = self.payload_index.borrow();
+                let filtered_points = payload_index.query_points(filter);
```

Comment:
now we re-evaluate the filtered points for each `query_vector` driven by the loop in the caller